File: node_backend/src/controllers/user_controller.ts
// controllers/user_controller.ts — Equivalent to go_backend/controllers/user_controller.go
import { Request, Response } from 'express';
import { UserService, CreateUserRequest } from '../services/user_service';
import * as apiResponse from '../utils/response';

/**
 * UserController — handles HTTP requests for user operations.
 * Mirrors the Go UserController struct with dependency injection.
 */
export class UserController {
    private userService: UserService;

    constructor(userService: UserService) {
        this.userService = userService;
    }

    /**
     * POST /api/users — Create a new user.
     */
    createUser = async (req: Request, res: Response): Promise<void> => {
        try {
            const { name, email } = req.body as CreateUserRequest;

            if (!name || !email) {
                apiResponse.error(res, 400, 'Name and email are required');
                return;
            }

            // Basic email validation
            const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
            if (!emailRegex.test(email)) {
                apiResponse.error(res, 400, 'Invalid email format');
                return;
            }

            const user = await this.userService.createUser({ name, email });
            apiResponse.success(res, 201, 'User created successfully', user);
        } catch (err) {
            const message = err instanceof Error ? err.message : 'Internal Server Error';
            apiResponse.error(res, 500, message);
        }
    };

    /**
     * GET /api/users/:id — Get a user by ID.
     */
    getUser = async (req: Request, res: Response): Promise<void> => {
        try {
            const id = parseInt(req.params.id, 10);
            if (isNaN(id)) {
                apiResponse.error(res, 400, 'Invalid user ID');
                return;
            }

            const user = await this.userService.getUser(id);
            apiResponse.success(res, 200, 'User retrieved successfully', user);
        } catch (err) {
            apiResponse.error(res, 404, 'User not found');
        }
    };
}

File: go_backend/controllers/user_controller.go
package controllers

import (
	"net/http"
	"strconv"

	"syncflow/services"
	"syncflow/utils"
	"github.com/gin-gonic/gin"
)

type UserController struct {
	userService services.UserService
}

func NewUserController(userService services.UserService) *UserController {
	return &UserController{userService: userService}
}

func (ctrl *UserController) CreateUser(c *gin.Context) {
	var req services.CreateUserRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.Error(c, http.StatusBadRequest, err.Error())
		return
	}

	user, err := ctrl.userService.CreateUser(c.Request.Context(), req)
	if err != nil {
		utils.Error(c, http.StatusInternalServerError, err.Error())
		return
	}

	utils.Success(c, http.StatusCreated, "User created successfully", user)
}

func (ctrl *UserController) GetUser(c *gin.Context) {
	idStr := c.Param("id")
	id, err := strconv.ParseUint(idStr, 10, 32)
	if err != nil {
		utils.Error(c, http.StatusBadRequest, "Invalid user ID")
		return
	}

	user, err := ctrl.userService.GetUser(c.Request.Context(), uint(id))
	if err != nil {
		utils.Error(c, http.StatusNotFound, "User not found")
		return
	}

	utils.Success(c, http.StatusOK, "User retrieved successfully", user)
}
